File: runner/runner.go
// Package runner is the harness every engine written in Go is measured by.
//
// The point of putting it here rather than in each runner is that the timing,
// the batching and the phase boundaries are then provably the same for every
// engine. A benchmark where each subject brings its own stopwatch measures the
// stopwatches.
//
// An engine written in another language reimplements this, and the contract it
// has to match is written down in [github.com/tamnd/kura-bench/bench] rather
// than only expressed in this code.
package runner

import (
	"bufio"
	"context"
	"errors"
	"flag"
	"fmt"
	"os"
	"runtime/debug"
	"strings"
	"sync"
	"time"

	"github.com/tamnd/kura-bench/bench"
	"github.com/tamnd/kura-bench/corpus"
)

// BatchSize is how many documents are handed to an engine at once.
//
// It is the same for every engine on purpose. Each of them has a batch size its
// own documentation recommends, and using each one's favourite would measure
// the recommendations rather than the engines. Five hundred is large enough
// that per call overhead has stopped mattering for all of them and small enough
// that none of them is holding the corpus in memory to satisfy it.
const BatchSize = 500

// Info is what an engine calls itself.
type Info struct {
	Name     string
	Version  string
	Language string
}

// Engine is one search engine under test.
//
// The lifecycle is Create, then AddBatch until the corpus is exhausted, then
// Flush, then Close. A second process calls Open and then Search.
//
// Search must be safe to call from several goroutines at once, because the
// throughput phase does. An engine that cannot do that should say so in its
// notes and the harness will report the serial numbers only.
type Engine interface {
	Describe() Info

	// Create makes an empty index in dir. The directory exists and is empty.
	Create(dir string) error

	// AddBatch indexes documents. It may return before they are queryable.
	AddBatch(docs []corpus.Document) error

	// Flush makes everything added queryable and durable. The indexing phase is
	// timed to the end of this call, not to the end of the last AddBatch,
	// because an engine that returns early from writes and does the work in the
	// background has not done less work.
	Flush() error

	// Open attaches to an index already on disk.
	Open(dir string) error

	// Search runs one query and returns how many documents matched in total,
	// which is not the same as how many were returned.
	Search(ctx context.Context, query string, limit int) (int, error)

	// Close releases everything.
	Close() error
}

// ModuleVersion returns the version of a dependency as it was actually built,
// which is the only version worth printing in a result. A constant in the
// source says what somebody meant to use.
func ModuleVersion(path string) string {
	info, ok := debug.ReadBuildInfo()
	if !ok {
		return "unknown"
	}
	for _, d := range info.Deps {
		if d.Path == path {
			if d.Replace != nil {
				return d.Replace.Version + " (replaced)"
			}
			return d.Version
		}
	}
	return "unknown"
}

// Config is the parsed command line.
type Config struct {
	Corpus  string
	Queries string
	Work    string
	Phase   string
	Repeat  int
	Limit   int
	Workers int
}

// Main is the whole of a runner's main function.
//
// It parses the standard flags, runs the phase it was asked for, and writes one
// JSON object to standard output. Anything it wants to say to a person goes to
// standard error, because standard output is the result and nothing else.
func Main(newEngine func(Config) (Engine, error)) {
	cfg := Config{}
	flag.StringVar(&cfg.Corpus, "corpus", "", "path to the corpus jsonl file")
	flag.StringVar(&cfg.Queries, "queries", "", "path to the query file, one per line")
	flag.StringVar(&cfg.Work, "work", "", "directory the index is built in")
	flag.StringVar(&cfg.Phase, "phase", "all", "index, query or all")
	flag.IntVar(&cfg.Repeat, "repeat", 20, "how many times each query is timed")
	flag.IntVar(&cfg.Limit, "limit", 0, "stop after this many documents, zero for all")
	flag.IntVar(&cfg.Workers, "workers", 0, "queries in flight for the throughput phase, zero for one per core")
	flag.Parse()

	if err := run(cfg, newEngine); err != nil {
		fmt.Fprintln(os.Stderr, "runner:", err)
		os.Exit(1)
	}
}

func run(cfg Config, newEngine func(Config) (Engine, error)) error {
	if cfg.Corpus == "" || cfg.Work == "" {
		return errors.New("both -corpus and -work are required")
	}
	eng, err := newEngine(cfg)
	if err != nil {
		return err
	}
	info := eng.Describe()

	res := bench.Result{
		Engine:   info.Name,
		Version:  info.Version,
		Language: info.Language,
		Machine:  bench.Describe(),
	}

	switch cfg.Phase {
	case "index":
		if err := indexPhase(cfg, eng, &res); err != nil {
			return err
		}
	case "query":
		if err := queryPhase(cfg, eng, &res); err != nil {
			return err
		}
	case "all":
		if err := indexPhase(cfg, eng, &res); err != nil {
			return err
		}
		if err := eng.Close(); err != nil {
			return err
		}
		// Reopening in the same process is not a cold start, and the result
		// says so. A real one needs a second process, which is what the
		// orchestrator does when it runs the two phases separately.
		res.Notes = strings.TrimSpace(res.Notes +
			" the open phase ran in the same process as the build, so it is warmer than a real cold start")
		fresh, err := newEngine(cfg)
		if err != nil {
			return err
		}
		eng = fresh
		if err := queryPhase(cfg, eng, &res); err != nil {
			return err
		}
	default:
		return fmt.Errorf("unknown phase %q", cfg.Phase)
	}

	if err := eng.Close(); err != nil {
		return err
	}
	return res.Write(os.Stdout)
}

// indexPhase builds the index from empty and times it to the end of the flush.
func indexPhase(cfg Config, eng Engine, res *bench.Result) error {
	if err := os.MkdirAll(cfg.Work, 0o755); err != nil {
		return err
	}
	if err := eng.Create(cfg.Work); err != nil {
		return fmt.Errorf("create: %w", err)
	}

	batch := make([]corpus.Document, 0, BatchSize)
	var stats bench.CorpusStats

	start := bench.Take()
	read, err := corpus.ReadFile(cfg.Corpus, func(d corpus.Document) error {
		if cfg.Limit > 0 && stats.Documents >= cfg.Limit {
			return corpus.ErrStop
		}
		stats.Documents++
		stats.Bytes += int64(len(d.Body))
		batch = append(batch, d)
		if len(batch) < BatchSize {
			return nil
		}
		if err := eng.AddBatch(batch); err != nil {
			return err
		}
		batch = batch[:0]
		return nil
	})
	if err != nil && !errors.Is(err, corpus.ErrStop) {
		return fmt.Errorf("indexing: %w", err)
	}
	if len(batch) > 0 {
		if err := eng.AddBatch(batch); err != nil {
			return fmt.Errorf("indexing the last batch: %w", err)
		}
	}
	if err := eng.Flush(); err != nil {
		return fmt.Errorf("flush: %w", err)
	}
	usage := bench.Measure(start)

	if cfg.Limit == 0 {
		stats = bench.CorpusStats{Documents: read.Documents, Bytes: read.Bytes}
	}
	size, files, err := bench.DirSize(cfg.Work)
	if err != nil {
		return err
	}

	res.Corpus = stats
	res.Index = bench.IndexPhase{Usage: usage, Bytes: size, Files: files}
	fmt.Fprintf(os.Stderr, "indexed %d documents in %.1fs, %.1f MB/s, index %.1f MB\n",
		stats.Documents, usage.WallSeconds, res.MBPerSecond(), float64(size)/(1<<20))
	return nil
}

// queryPhase opens an index already on disk and runs the query set.
func queryPhase(cfg Config, eng Engine, res *bench.Result) error {
	queries, err := readQueries(cfg.Queries)
	if err != nil {
		return err
	}
	ctx := context.Background()

	// Opening and answering one query is the cold start, and the one query is
	// part of it on purpose. Several of these engines do almost nothing in
	// their open call and pay for it on the first search instead, and an open
	// timed without a query would report that as free.
	openStart := bench.Take()
	if err := eng.Open(cfg.Work); err != nil {
		return fmt.Errorf("open: %w", err)
	}
	if _, err := eng.Search(ctx, queries[0], 10); err != nil {
		return fmt.Errorf("the first query after open: %w", err)
	}
	openUsage := bench.Measure(openStart)
	res.Open = bench.OpenPhase{Usage: openUsage, ResidentBytes: openUsage.RSSBytes}

	searchStart := bench.Take()
	stats := make([]bench.QueryStat, 0, len(queries))
	for _, q := range queries {
		// One warm up that is not counted, because the first run of a query
		// pays for whatever the engine caches per term and no deployment sees
		// that cost on every request.
		hits, err := eng.Search(ctx, q, 10)
		if err != nil {
			return fmt.Errorf("query %q: %w", q, err)
		}
		runs := make([]time.Duration, 0, cfg.Repeat)
		for range cfg.Repeat {
			t := time.Now()
			hits, err = eng.Search(ctx, q, 10)
			if err != nil {
				return fmt.Errorf("query %q: %w", q, err)
			}
			runs = append(runs, time.Since(t))
		}
		stat := bench.Summarise(q, hits, runs)
		stats = append(stats, stat)
		// One line per query rather than per run, so a slow engine says where
		// it has got to instead of looking hung for an hour. It is printed
		// after the timed runs and never between them.
		fmt.Fprintf(os.Stderr, "  %-40s %8d hits  %s median\n",
			q, stat.Hits, time.Duration(stat.MedianMS*float64(time.Millisecond)).Round(time.Microsecond))
	}
	searchUsage := bench.Measure(searchStart)

	res.Search = bench.SearchPhase{Usage: searchUsage, Queries: stats}
	res.Search.Concurrent = concurrent(ctx, eng, queries, cfg)
	res.Update = updatePhase(cfg, eng, res)
	return nil
}

// UpdateDocuments is how many documents the update phase rewrites.
//
// It is a fixed count rather than a fraction of the corpus so that the figure
// means the same thing on a machine that ran the whole corpus and on one that
// was given a tenth of it. Five thousand is roughly a busy day of edits on a
// repository this size, and it is enough work that an engine which handles
// updates by rebuilding is obvious rather than merely slow.
const UpdateDocuments = 5000

// updatePhase reindexes a slice of the corpus into the index that is already
// open, which is what an incremental sync does and is not the same operation as
// building from empty.
//
// An engine that cannot write to an index it opened for reading returns an
// error here, and the phase is left out with the reason in the notes instead of
// being reported as a zero.
func updatePhase(cfg Config, eng Engine, res *bench.Result) *bench.UpdatePhase {
	want := UpdateDocuments
	if cfg.Limit > 0 && cfg.Limit < want {
		// Rewriting documents the index does not have would be an insert
		// measured as an update, which is a different and easier operation.
		want = cfg.Limit
	}

	batch := make([]corpus.Document, 0, BatchSize)
	var (
		done  int
		bytes int64
		bad   error
	)

	start := bench.Take()
	_, err := corpus.ReadFile(cfg.Corpus, func(d corpus.Document) error {
		if done >= want {
			return corpus.ErrStop
		}
		done++
		bytes += int64(len(d.Body))
		batch = append(batch, d)
		if len(batch) < BatchSize {
			return nil
		}
		if err := eng.AddBatch(batch); err != nil {
			bad = err
			return corpus.ErrStop
		}
		batch = batch[:0]
		return nil
	})
	if err != nil && !errors.Is(err, corpus.ErrStop) {
		bad = err
	}
	if bad == nil && len(batch) > 0 {
		bad = eng.AddBatch(batch)
	}
	if bad == nil {
		bad = eng.Flush()
	}
	usage := bench.Measure(start)

	if bad != nil {
		res.Notes = strings.TrimSpace(res.Notes +
			" the update phase was left out because this engine cannot reindex into an index it opened: " + bad.Error())
		return nil
	}

	size, _, err := bench.DirSize(cfg.Work)
	if err != nil {
		size = 0
	}
	return &bench.UpdatePhase{
		Usage:           usage,
		Documents:       done,
		Bytes:           bytes,
		IndexBytesAfter: size,
	}
}

// concurrent runs the query set with several in flight, which is the only way
// to get a throughput number that means anything. Dividing one second by the
// serial latency gives a figure no deployment has ever reached.
func concurrent(ctx context.Context, eng Engine, queries []string, cfg Config) *bench.ConcurrentStat {
	workers := cfg.Workers
	if workers <= 0 {
		workers = len(queries)
	}
	if workers > 64 {
		workers = 64
	}
	total := len(queries) * cfg.Repeat

	jobs := make(chan string, workers)
	times := make([]time.Duration, total)
	var (
		wg  sync.WaitGroup
		mu  sync.Mutex
		bad error
		i   int
	)

	start := time.Now()
	for range workers {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for q := range jobs {
				t := time.Now()
				_, err := eng.Search(ctx, q, 10)
				d := time.Since(t)
				mu.Lock()
				if err != nil && bad == nil {
					bad = err
				}
				if i < len(times) {
					times[i] = d
					i++
				}
				mu.Unlock()
			}
		}()
	}
	for range cfg.Repeat {
		for _, q := range queries {
			jobs <- q
		}
	}
	close(jobs)
	wg.Wait()
	elapsed := time.Since(start)

	if bad != nil {
		// An engine whose Search is not safe for concurrent use fails here, and
		// the throughput figure is left out rather than reported as whatever
		// the race happened to produce.
		fmt.Fprintf(os.Stderr, "concurrent phase failed, leaving it out: %v\n", bad)
		return nil
	}

	summary := bench.Summarise("", 0, times[:i])
	return &bench.ConcurrentStat{
		Workers:  workers,
		Queries:  i,
		Seconds:  elapsed.Seconds(),
		MedianMS: summary.MedianMS,
		P99MS:    summary.P99MS,
	}
}

// readQueries reads one query per line, ignoring blanks and comments.
func readQueries(name string) ([]string, error) {
	if name == "" {
		return nil, errors.New("-queries is required for the query phase")
	}
	f, err := os.Open(name)
	if err != nil {
		return nil, err
	}
	defer func() { _ = f.Close() }()

	var out []string
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		out = append(out, line)
	}
	if err := sc.Err(); err != nil {
		return nil, err
	}
	if len(out) == 0 {
		return nil, fmt.Errorf("%s has no queries in it", name)
	}
	return out, nil
}

File: runners/java/lucene/Runner.java
// Measures Lucene against the same corpus every other engine here is given.
//
// Lucene is in this table because it is what most enterprise search actually
// runs on. Elasticsearch and OpenSearch are Lucene with a cluster around them,
// and a comparison that leaves it out is measuring against the rivals that are
// convenient rather than the one people have deployed.
//
// The batch size, the phase boundaries and the timing are the ones the shared
// harness uses, because a benchmark where each subject brings its own stopwatch
// measures the stopwatches. The measuring half is in Bench.java next to this,
// so this file is an engine and nothing else.
//
// Two things about the Java Virtual Machine belong in the reader's head before
// the memory column does. Peak resident memory includes heap the collector has
// not handed back, so it is the memory the process asked for and not the memory
// it needed. And the code is interpreted until it is compiled, so the first
// runs of a phase are slower than the rest. Neither is a flaw in Lucene and
// both are what a deployment actually pays, which is why they are reported as
// they are rather than tuned away.

import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.CharArraySet;
import org.apache.lucene.analysis.standard.StandardAnalyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.StoredField;
import org.apache.lucene.document.StringField;
import org.apache.lucene.document.TextField;
import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.index.IndexWriter;
import org.apache.lucene.index.IndexWriterConfig;
import org.apache.lucene.index.StoredFields;
import org.apache.lucene.index.Term;
import org.apache.lucene.queryparser.classic.MultiFieldQueryParser;
import org.apache.lucene.queryparser.classic.ParseException;
import org.apache.lucene.queryparser.classic.QueryParser;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.search.TopDocs;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.FSDirectory;
import org.apache.lucene.util.Version;

public final class Runner {
    private Runner() {}

    /**
     * The writer's memory budget before it flushes a segment. It is the single
     * setting that most changes the shape of the index, and it is the same two
     * hundred megabytes the Tantivy runner is given, so that the index phase
     * compares the two engines rather than comparing two numbers somebody
     * picked. Lucene's own default is sixteen, which is a figure for a machine
     * indexing alongside everything else it is doing and not one for a bulk
     * load.
     */
    private static final double WRITER_HEAP_MB = 200;

    // The fields searched by a bare query, which is title and body here for
    // the same reason it is on every other engine in this suite.
    private static final String[] SEARCHED = {"title", "body"};

    public static void main(String[] args) {
        try {
            run(args);
        } catch (Exception e) {
            System.err.println("lucene-runner: " + e);
            e.printStackTrace(System.err);
            System.exit(1);
        }
    }

    private static void run(String[] args) throws IOException {
        Bench.Config cfg = Bench.parse(args);
        Bench.Result res = new Bench.Result();
        res.engine = "lucene";
        // Lucene's version and not this file's. A number without the engine's
        // own version is not reproducible.
        res.version = Version.LATEST.toString();
        res.language = "java";
        res.machine = Bench.describe();
        res.notes = "peak resident memory is the whole virtual machine, so it includes heap the "
                + "collector has not returned and is the memory the process asked for rather "
                + "than the memory it needed. Documents are fed to the writer from one thread, "
                + "the way the harness feeds every other engine, so the index phase does not "
                + "use the several threads a Lucene bulk load usually gets.";

        switch (cfg.phase) {
            case "index" -> indexPhase(cfg, res);
            case "query" -> queryPhase(cfg, res);
            case "all" -> {
                indexPhase(cfg, res);
                res.notes = "the open phase ran in the same process as the build, so it is "
                        + "warmer than a real cold start. " + res.notes;
                queryPhase(cfg, res);
            }
            default -> throw new IllegalArgumentException("unknown phase " + cfg.phase);
        }

        System.out.println(Bench.json(res));
    }

    /**
     * The analyzer, spelled out rather than defaulted.
     *
     * Lowercasing and no stopwords is what the other engines here do, and
     * naming the empty set means a change in what the no argument constructor
     * defaults to cannot quietly change what is being compared. No stemming,
     * because none of the others stem either.
     */
    private static Analyzer analyzer() {
        return new StandardAnalyzer(CharArraySet.EMPTY_SET);
    }

    /**
     * The document, stored as well as indexed.
     *
     * Every engine in this comparison is asked to be able to show a result to a
     * person, and one that keeps nothing would look smaller on disk for a
     * reason that has nothing to do with the index.
     */
    private static Document document(Bench.Document d) {
        Document doc = new Document();
        doc.add(new StringField("id", d.id, Field.Store.YES));
        doc.add(new StringField("repo", d.repo, Field.Store.YES));
        // The path is worth showing and not worth searching as one term, so it
        // is stored and not indexed.
        doc.add(new StoredField("path", d.path));
        doc.add(new TextField("title", d.title, Field.Store.YES));
        doc.add(new TextField("body", d.body, Field.Store.YES));
        doc.add(new StringField("ext", d.ext, Field.Store.YES));
        return doc;
    }

    private static IndexWriterConfig writerConfig(IndexWriterConfig.OpenMode mode) {
        IndexWriterConfig conf = new IndexWriterConfig(analyzer());
        conf.setOpenMode(mode);
        conf.setRAMBufferSizeMB(WRITER_HEAP_MB);
        return conf;
    }

    private static void indexPhase(Bench.Config cfg, Bench.Result res) throws IOException {
        Files.createDirectories(cfg.work);

        int[] documents = {0};
        long[] bytes = {0};
        Bench.Snapshot start = Bench.take();
        try (Directory dir = FSDirectory.open(cfg.work);
                IndexWriter writer = new IndexWriter(dir, writerConfig(IndexWriterConfig.OpenMode.CREATE))) {
            Bench.read(cfg.corpus, d -> {
                if (cfg.limit > 0 && documents[0] >= cfg.limit) {
                    return false;
                }
                documents[0]++;
                bytes[0] += Bench.utf8Length(d.body);
                try {
                    writer.addDocument(document(d));
                } catch (IOException e) {
                    throw new UncheckedWrite(e);
                }
                return true;
            });
            // The phase is timed to the end of the commit and the close, not
            // to the end of the last add, because an engine that returns early
            // from writes and finishes the merge in the background has not
            // done less work. Closing the writer is what waits for it.
            writer.commit();
        } catch (UncheckedWrite e) {
            throw e.cause();
        }
        Bench.Usage phase = Bench.measure(start);

        long[] size = Bench.dirSize(cfg.work);
        res.documents = documents[0];
        res.corpusBytes = bytes[0];
        res.indexUsage = phase;
        res.indexBytes = size[0];
        res.indexFiles = (int) size[1];
        System.err.printf(
                "indexed %d documents in %.1fs, %.1f MB/s, index %.1f MB%n",
                documents[0],
                phase.wallSeconds,
                bytes[0] / (double) (1 << 20) / phase.wallSeconds,
                size[0] / (double) (1 << 20));
    }

    private static void queryPhase(Bench.Config cfg, Bench.Result res) throws IOException {
        List<String> queries = Bench.queries(cfg.queries);
        if (queries.isEmpty()) {
            throw new IllegalArgumentException("the query file has no queries in it");
        }

        // Opening and answering one query is the cold start, and the one query
        // is part of it on purpose. Lucene does very little in open and pays
        // for it on the first search, and an open timed without a query would
        // report that as free.
        Bench.Snapshot openStart = Bench.take();
        Directory dir = FSDirectory.open(cfg.work);
        DirectoryReader reader = DirectoryReader.open(dir);
        // No executor, so a query runs on the thread that asked for it. That is
        // what the other engines here do, and handing Lucene a pool would make
        // its latency column describe a different amount of hardware.
        IndexSearcher searcher = new IndexSearcher(reader);
        QueryParser parser = parser();
        searchOnce(searcher, parser, queries.get(0), null);
        Bench.Usage open = Bench.measure(openStart);
        res.openUsage = open;
        res.openResidentBytes = open.rssBytes;

        Bench.Snapshot searchStart = Bench.take();
        List<Bench.QueryStat> stats = new ArrayList<>(queries.size());
        List<String> ids = new ArrayList<>(Bench.SEARCH_LIMIT);
        for (String q : queries) {
            // One warm up that is not counted, because the first run of a
            // query pays for whatever the engine caches per term and no
            // deployment sees that cost on every request.
            //
            // The page comes off this run for the same reason. Keeping the
            // identifiers allocates, and the run that pays for it is the one
            // whose time nobody reads.
            ids.clear();
            int hits = searchOnce(searcher, parser, q, ids);
            List<Double> runs = new ArrayList<>(cfg.repeat);
            for (int i = 0; i < cfg.repeat; i++) {
                long t = System.nanoTime();
                hits = searchOnce(searcher, parser, q, null);
                runs.add((System.nanoTime() - t) / 1e6);
            }
            Bench.QueryStat stat = Bench.summarise(q, hits, runs);
            stat.ids.addAll(ids);
            stats.add(stat);
        }
        Bench.Usage search = Bench.measure(searchStart);

        res.searchUsage = search;
        res.queries = stats;
        res.concurrent = concurrentPhase(searcher, queries, cfg);

        reader.close();
        dir.close();
        updatePhase(cfg, res);
    }

    private static QueryParser parser() {
        MultiFieldQueryParser p = new MultiFieldQueryParser(SEARCHED, analyzer());
        // A bare query is read as OR, which is how every other engine here is
        // asked to read it. Lucene defaults to that already and it is set
        // explicitly so that a change in its default does not silently change
        // what is being compared.
        p.setDefaultOperator(QueryParser.Operator.OR);
        return p;
    }

    /**
     * Runs one query and returns the total number of matches, which is not the
     * same as the number returned. Both the count and the page are asked for,
     * because the other engines here report both and dropping one would make
     * the numbers describe different work.
     *
     * When ids is given it collects the identifiers of the page. The documents
     * are fetched either way, so the only extra work is reading a field out of
     * one that is already in hand.
     */
    private static int searchOnce(
            IndexSearcher searcher, QueryParser parser, String text, List<String> ids)
            throws IOException {
        Query query = parse(parser, text);
        int count = searcher.count(query);
        TopDocs top = searcher.search(query, Bench.SEARCH_LIMIT);
        StoredFields stored = searcher.storedFields();
        for (ScoreDoc hit : top.scoreDocs) {
            Document doc = stored.document(hit.doc);
            if (ids != null) {
                String id = doc.get("id");
                if (id != null) {
                    ids.add(id);
                }
            }
        }
        return count;
    }

    // The query set is plain terms, but a corpus derived set can turn up a
    // token the parser treats as syntax. Tantivy's parser is lenient about
    // that and Lucene's is not, so a query it refuses is retried as literal
    // text rather than failing the run or being quietly dropped.
    private static Query parse(QueryParser parser, String text) {
        try {
            return parser.parse(text);
        } catch (ParseException e) {
            try {
                return parser.parse(QueryParser.escape(text));
            } catch (ParseException fatal) {
                throw new IllegalArgumentException("cannot parse query " + text, fatal);
            }
        }
    }

    /**
     * The query set run with several in flight, which is the only way to get a
     * throughput number that means anything. Dividing one second by the serial
     * latency gives a figure no deployment has ever reached.
     */
    private static Bench.ConcurrentStat concurrentPhase(
            IndexSearcher searcher, List<String> queries, Bench.Config cfg) {
        int workers = cfg.workers == 0 ? queries.size() : cfg.workers;
        workers = Math.clamp(workers, 1, 64);

        List<String> jobs = Bench.jobs(queries, cfg.repeat);
        AtomicInteger next = new AtomicInteger();
        List<List<Double>> collected = new ArrayList<>(workers);
        List<Thread> threads = new ArrayList<>(workers);
        boolean[] failed = {false};

        long start = System.nanoTime();
        for (int w = 0; w < workers; w++) {
            List<Double> mine = new ArrayList<>();
            collected.add(mine);
            // Every worker parses for itself. The parser holds state while it
            // is working and sharing one would serialise the phase that exists
            // to measure what happens when nothing is serialised.
            QueryParser parser = parser();
            Thread t = new Thread(() -> {
                while (true) {
                    int i = next.getAndIncrement();
                    if (i >= jobs.size()) {
                        return;
                    }
                    long at = System.nanoTime();
                    try {
                        searchOnce(searcher, parser, jobs.get(i), null);
                    } catch (IOException | RuntimeException e) {
                        failed[0] = true;
                        return;
                    }
                    mine.add((System.nanoTime() - at) / 1e6);
                }
            });
            threads.add(t);
            t.start();
        }
        for (Thread t : threads) {
            try {
                t.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return null;
            }
        }
        double elapsed = (System.nanoTime() - start) / 1e9;
        if (failed[0]) {
            // A worker that failed means the throughput figure would describe
            // whatever the failure happened to produce, so it is left out.
            return null;
        }

        List<Double> all = new ArrayList<>();
        for (List<Double> some : collected) {
            all.addAll(some);
        }
        Bench.QueryStat summary = Bench.summarise("", 0, all);
        Bench.ConcurrentStat stat = new Bench.ConcurrentStat();
        stat.workers = workers;
        stat.queries = all.size();
        stat.seconds = elapsed;
        stat.medianMs = summary.medianMs;
        stat.p99Ms = summary.p99Ms;
        return stat;
    }

    /**
     * Reindexes a slice of the corpus into the index that is already built,
     * which is what an incremental sync does and is not the same operation as
     * building from empty.
     */
    private static void updatePhase(Bench.Config cfg, Bench.Result res) throws IOException {
        int want = cfg.capped(Bench.UPDATE_DOCUMENTS);

        int[] documents = {0};
        long[] bytes = {0};
        Bench.Snapshot start = Bench.take();
        try (Directory dir = FSDirectory.open(cfg.work);
                IndexWriter writer = new IndexWriter(dir, writerConfig(IndexWriterConfig.OpenMode.APPEND))) {
            Bench.read(cfg.corpus, d -> {
                if (documents[0] >= want) {
                    return false;
                }
                documents[0]++;
                bytes[0] += Bench.utf8Length(d.body);
                try {
                    // Replacing by term is what an update is. Adding without
                    // it would double the corpus and report a rate for an
                    // operation nobody runs.
                    writer.updateDocument(new Term("id", d.id), document(d));
                } catch (IOException e) {
                    throw new UncheckedWrite(e);
                }
                return true;
            });
            writer.commit();
        } catch (UncheckedWrite e) {
            throw e.cause();
        }
        Bench.Usage usage = Bench.measure(start);

        long[] size = Bench.dirSize(cfg.work);
        res.updateUsage = usage;
        res.updateDocuments = documents[0];
        res.updateBytes = bytes[0];
        res.updateIndexBytesAfter = size[0];
    }

    /** Carries a write failure out of the corpus callback, which cannot throw
     * a checked exception, without losing it. */
    private static final class UncheckedWrite extends RuntimeException {
        private static final long serialVersionUID = 1L;

        UncheckedWrite(IOException cause) {
            super(cause);
        }

        IOException cause() {
            return (IOException) getCause();
        }
    }
}
